src/test/tcp/test_tcp.c

```diff
@@ -782,9 +844,17 @@ int main(int argc, char *argv[]) {
             MYLOG("error, client mode also needs a server ip address; see usage");
             return -1;
         }
+        if (argc == 5) {
+            queuename = argv[4];
```

Comment:
We should make this argument mandatory rather than having a default. We want to be careful to avoid diffiult-to-debug errors in the case that multiple concurrent tests accidentally try to use the same queue name

Reply:
To do that I should provide a unique filename for each test. Which files can I take ?

Reply:
We can use an empty file named after the test-name. e.g. something like (pseudo-code; might need some escaping to embed a shell variable)

```
add_test(
    NAME tcp-blocking-loopback
- COMMAND shadow-test-launcher test-tcp blocking ./test-tcp server : test-tcp blocking ./test-tcp client localhost
+ COMMAND /bin/bash -c "QFILE=tcp-blocking-loopback-qfile && shadow-test-launcher test-tcp blocking ./test-tcp server $QFILE : test-tcp blocking ./test-tcp client localhost $QFILE && rm $QFILE"
```

It'd be nice if `ctest` set some environment variable to the test-name, in which case we could just use that instead of having to pipe it through ourselves, but as far as I can tell it doesn't.

Reply:
Actually, since this can be a regular file, we can just use `mktemp` to get us a guaranteed-unique filename in the `tmp` directory. e.g.

```
COMMAND /bin/bash -c 'QFILE=`mktemp` && ...
```

Reply:
Alternatively, since we're forced to handle this in the wrapper script anyways, we could use `ipcmk` to generate a distinct message queue for each test, and pass the generated key directly instead of using ftok.

Reply:
The only problem with one of those technics is the difficulty to provide dynamic arguments when we run tests with Shadow. I will let you take a look at what I did. I hope there is a better and simpler way !

Reply:
Yes, I'd overlooked the difficulty of passing a dynamic command-line argument to Shadow. I don't know of an easy way to do this, currently.

I was surprised what you did seemed to work; I checked it out locally and found that actually the tests are failing, but the failures are being suppressed. In:

```
/bin/bash -c "QUEUE=`ipcmk --queue | sed 's/[^0-9]*//g'` && ${CMAKE_BINARY_DIR}/src/main/shadow -l ...; ipcrm -q $QUEUE"
```

The bash process ends up returning the exit code of `ipcrm` rather than shadow's exit code. If we wanted to ensure cleanup while returning the correct exit code it'd need to be something like:

```
/bin/bash -c "QUEUE=`ipcmk --queue | sed 's/[^0-9]*//g'` && ${CMAKE_BINARY_DIR}/src/main/shadow -l ...; RV=$?; ipcrm -q $QUEUE; exit $RV"
```

As for dealing with the dynamic argument issue... the cleanest workaround I can think of is to use an environment variable instead of a command-line argument. Then you can grab it from inside the program using `getenv`. You'd need to add an `export` to the shell snippet. e.g.

```
/bin/bash -c "export QUEUE=`ipcmk --queue | sed 's/[^0-9]*//g'` && ${CMAKE_BINARY_DIR}/src/main/shadow -l ...; RV=$?; ipcrm -q $QUEUE; exit $RV"
```

Perhaps it'd be better to encapsulate this in a small shell script. I think this should do it:

```
#!/bin/bash

# Catch failures
set -euo pipefail

# Create Q, and export into environment of child processes
export QUEUE=`ipcmk --queue | sed 's/[^0-9]*//g'`

# Ensure Q is removed on exit
trap "ipcrm -q $QUEUE" EXIT

# Interpret remaining args as a command to run:
$@
```

The test commands could then be something like:
```
COMMAND ${CMAKE_SOURCE_DIR}/with_q.sh ${CMAKE_BINARY_DIR}/src/main/shadow -l debug -d blocking-loopback.shadow.data ${CMAKE_CURRENT_SOURCE_DIR}/tcp-blocking-loopback.test.shadow.config.xml
```

Reply:
Creating the queue from script seem ok, but not destructing. Even with the atexit method, I see queues in my IPC after tests.